### ISSUE TITLE
Better Hyperparameter Searching

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#254]
The current `HyperparamOpt` class accepts lists of values for each hyperparameter, and then searches over the product space. For example
- learning rate: `.001`, `.0003`
- hidden layer size: `100`, `1000`

Would generate 4 combinations of hyperparameters `(.001, 100), (.001, 1000), (.0003, 100), (.0003, 1000)`. This type of product expansion is not always desirable. We might for example want to try just `(.001, 100), (.0003, 1000)`. (This particular example doesn't make much sense, but it is common for groups of hyperparameters to be tied together). Allowing the user to specify tied hyperparameter settings, in a list, would be a nice complement to the product based search currently supported.

**Comment by joegomes** [comment]
@rbharath I think is a good idea.  It would be nice to be able to distribute model fitting jobs spawned by hyperparameter opt to multiple GPUs with a foreman/worker queue.

**Comment by rbharath** [comment]
Yes definitely. One sticking point is that different cluster environments generally tend to have different needs. It might be best to move hyperparameter search into another repo entirely, say deepchem/deepchem-hyp and offer explicit support for workqueue integration there (like ForceBalance does).

**Comment by joegomes** [comment]
I agree with the deepchem-hyp idea, potentially a very useful tool and would be best to not further complicate that code/deepchem installation.  I'll add it to my wishlist =)

**Comment by rbharath** [comment]
Some hyperparameter improvements have happened (gaussian process search for example), and the other specifics here don't seem to be a priority to closing. Feel free to open new issue for specific hyperparameter opt feature requests.
